sklearn/ensemble/_hist_gradient_boosting/tests/test_binning.py

```diff
@@ -312,3 +317,140 @@ def test_infinite_values():
+        (4, [2, 3, 3, 3, 3, 1, 3, 3, 0], [0, 7, 10]),
+        (5, [3, 1, 4, 4, 4, 2, 4, 4, 0], [0, 1, 7, 10]),
+        (6, [4, 1, 5, 5, 5, 3, 2, 5, 0], [0, 1, 4, 7, 10]),
+    ])
+def test_categorical_n_bins_less_than_cardinality(
+        n_bins, expected_trans, bin_categories):
+    # test when there is not enough bins only the top n_bins - 1 distinct
+    # values (based on cardinality) will have its own bin
+    X = np.array([[4] * 2 + [1] * 3 + [10] * 4 +
+                  [0] * 4 + [9] + [7] * 5], dtype=X_DTYPE).T
+    X_test = np.array([[10, 1, -1, 9, np.nan, 7, 4, 100, 0]], dtype=X_DTYPE).T
+
+    bin_mapper = _BinMapper(n_bins=n_bins,
+                            categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [n_bins - 1]
+    assert_allclose(bin_mapper.bin_categories_, [bin_categories])
+    X_trans = bin_mapper.transform(X_test)
+
+    # missing, negative, unknown values are mapped to the missing bin
+    expected_trans = np.array([expected_trans]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+def test_categorical_n_bins_less_than_cardinality_ties():
+    # when there are not enough bins and the cardiality is the same
+    # the elements with a higher lexicon ordering will be choosen as
+    # categories
+    X = np.array([[1] * 4 + [2] * 4 + [3] * 4 +
+                  [4] * 2 + [5] * 2 + [6] * 1], dtype=X_DTYPE).T
+    X_test = np.array([[1, 2, 3, 4, 5, 6, 7]], dtype=X_DTYPE).T
+
+    # With 4 bins used for non missing values. categories 1, 2, 3, 4
+    # will have its own bin, the rest will be placed in the missing bin.
+    bin_mapper = _BinMapper(n_bins=5, categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [4]
+    assert_allclose(bin_mapper.bin_categories_, [[1, 2, 3, 4]])
+
+    X_trans = bin_mapper.transform(X_test)
+    expected_trans = np.array([[0, 1, 2, 3, 4, 4, 4]]).T
+    assert_array_equal(X_trans, expected_trans)
+
+    # With 3 bins usd for non missing values. categoires 1, 2
+    # will have its own bin, the rest will be placed in the missing bin.
+    bin_mapper = _BinMapper(n_bins=3, categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [2]
+    assert_allclose(bin_mapper.bin_categories_, [[1, 2]])
+
+    X_trans = bin_mapper.transform(X_test)
+    expected_trans = np.array([[0, 1, 2, 2, 2, 2, 2]]).T
+    assert_array_equal(X_trans, expected_trans)
+
+
+@pytest.mark.parametrize("missing_value", [-1, np.nan])
+def test_categorical_default_categories_are_missing(missing_value):
+    # check that negative values are considered missing
+    X = np.array([[0] * 10 + [1] * 12 +
+                  [missing_value] * 13], dtype=X_DTYPE).T
+
+    bin_mapper = _BinMapper(n_bins=3, categorical=np.array([True])).fit(X)
+    assert bin_mapper.n_bins_non_missing_ == [2]
+    assert_array_equal(bin_mapper.bin_categories_, [[0, 1]])
+
+    X_test = np.array([[0, 1, missing_value, 2]], dtype=X_DTYPE).T
```

Comment:
Should we check both -1 and np.nan at test time, regardless of what was used fit?